src/main/java/org/opensearch/security/resources/ResourceAccessHandler.java

```diff
@@ -0,0 +1,506 @@
+import org.opensearch.security.spi.resources.sharing.Recipient;
+import org.opensearch.security.spi.resources.sharing.ResourceSharing;
+import org.opensearch.security.spi.resources.sharing.ShareWith;
+import org.opensearch.security.spi.resources.sharing.SharedWithActionGroup;
+import org.opensearch.security.support.ConfigConstants;
+import org.opensearch.security.user.User;
+import org.opensearch.threadpool.ThreadPool;
+
+/**
+ * This class handles resource access permissions for users, roles and backend-roles.
+ * It provides methods to check if a user has permission to access a resource
+ * based on the resource sharing configuration.
+ *
+ * @opensearch.experimental
+ */
+public class ResourceAccessHandler {
+    private static final Logger LOGGER = LogManager.getLogger(ResourceAccessHandler.class);
+
+    private final ThreadContext threadContext;
+    private final ResourceSharingIndexHandler resourceSharingIndexHandler;
+    private final AdminDNs adminDNs;
+    private final ResourcePluginInfo resourcePluginInfo;
+
+    public ResourceAccessHandler(
+        final ThreadPool threadPool,
+        final ResourceSharingIndexHandler resourceSharingIndexHandler,
+        AdminDNs adminDns,
+        ResourcePluginInfo resourcePluginInfo
+    ) {
+        this.threadContext = threadPool.getThreadContext();
+        this.resourceSharingIndexHandler = resourceSharingIndexHandler;
+        this.adminDNs = adminDns;
+        this.resourcePluginInfo = resourcePluginInfo;
+    }
+
+    /**
+     * Returns a set of accessible resource IDs for the current user within the specified resource index.
+     *
+     * @param resourceIndex The resource index to check for accessible resources.
+     * @param listener      The listener to be notified with the set of accessible resource IDs.
+     */
+    public void getAccessibleResourceIdsForCurrentUser(String resourceIndex, ActionListener<Set<String>> listener) {
+        final UserSubjectImpl userSubject = (UserSubjectImpl) threadContext.getPersistent(
+            ConfigConstants.OPENDISTRO_SECURITY_AUTHENTICATED_USER
+        );
+        final User user = (userSubject == null) ? null : userSubject.getUser();
+
+        // If no user is authenticated, return an empty set
+        if (user == null) {
+            LOGGER.warn("Unable to fetch user details. User is null.");
+            listener.onResponse(Collections.emptySet());
+            return;
+        }
+
+        LOGGER.debug("Listing accessible resources within the resource index {} for user: {}", resourceIndex, user.getName());
+
+        // 2. If the user is admin, simply fetch all resources
+        if (adminDNs.isAdmin(user)) {
+            loadAllResources(resourceIndex, ActionListener.wrap(listener::onResponse, listener::onFailure));
+            return;
+        }
+
+        // StepListener for the user’s "own" resources
+        StepListener<Set<String>> ownResourcesListener = new StepListener<>();
```

Comment:
Is it possible to do this in a single query instead of 4 separate queries?

Reply:
done.-